### PR TITLE
osd: additional protection for out-of-bounds EC reads

### PR DESCRIPTION
**Pull request by dillaman**
Fixes: http://tracker.ceph.com/issues/21629
Signed-off-by: Jason Dillaman <dillaman@redhat.com>

### CODE REVIEW

**Comment by liewegas** [comment]
Can we add a test case to test/librados/ that triggers this?

**Comment by dillaman** [comment]
@liewegas appended

File: src/osd/PrimaryLogPG.cc [hunk 1]
```diff
@@ -4618,13 +4618,28 @@ int PrimaryLogPG::do_extent_cmp(OpContext *ctx, OSDOp& osd_op)
   dout(20) << __func__ << dendl;
   ceph_osd_op& op = osd_op.op;
 
-  if (!ctx->obs->exists || ctx->obs->oi.is_whiteout()) {
+  auto& oi = ctx->new_obs.oi;
+  uint64_t size = oi.size;
+  if ((oi.truncate_seq < op.extent.truncate_seq) &&
+      (op.extent.offset + op.extent.length > op.extent.truncate_size)) {
+    size = op.extent.truncate_size;
+  }
```
Comment (liewegas) [inline src/osd/PrimaryLogPG.cc]: Does rbd actually use these truncation fields, or is this just here for completeness?

Comment (dillaman) [reply src/osd/PrimaryLogPG.cc]: No, RBD doesn't use it -- I added it to mimic "do_read"

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2017-10-04_21:58:33-rados-wip-sage-testing-2017-10-04-1404-distro-basic-smithi/
http://pulpito.ceph.com/sage-2017-10-05_02:36:25-rados-wip-sage-testing-2017-10-04-1404-distro-basic-smithi/

**Pull request merged**